db/db_query_plan.c

```diff
@@ -0,0 +1,197 @@
+/*
+   Copyright 2022 Bloomberg Finance L.P.
+
+   Licensed under the Apache License, Version 2.0 (the "License");
+   you may not use this file except in compliance with the License.
+   You may obtain a copy of the License at
+
+       http://www.apache.org/licenses/LICENSE-2.0
+
+   Unless required by applicable law or agreed to in writing, software
+   distributed under the License is distributed on an "AS IS" BASIS,
+   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+   See the License for the specific language governing permissions and
+   limitations under the License.
+*/
+
+#include "sql.h"
+
+hash_t *gbl_query_plan_hash = NULL;
+pthread_mutex_t gbl_query_plan_hash_mu = PTHREAD_MUTEX_INITIALIZER;
+int gbl_query_plan_max_queries = 1000;
+extern int gbl_debug_print_query_plans;
+extern double gbl_query_plan_percentage;
+
+static char *form_query_plan(const struct client_query_stats *query_stats)
+{
+    struct strbuf *query_plan_buf;
+    const struct client_query_path_component *c;
+    char *query_plan;
+
+    if (query_stats->n_components == 0) {
+        return NULL;
+    }
+
+    query_plan_buf = strbuf_new();
+    for (int i = 0; i < query_stats->n_components; i++) {
+        if (i > 0) {
+            strbuf_append(query_plan_buf, ", ");
+        }
+        c = &query_stats->path_stats[i];
+        strbuf_appendf(query_plan_buf, "table %s index %d", c->table, c->ix);
+    }
+
+    query_plan = strdup((char *)strbuf_buf(query_plan_buf));
+    strbuf_free(query_plan_buf);
+    return query_plan;
+}
+
+static void add_query_plan_int(const char *query, const char *query_plan, hash_t *query_plan_hash,
+                               int *alert_once_query, double current_cost_per_row)
+{
+    struct query_plan_item *q = hash_find(query_plan_hash, &query_plan);
+    if (q == NULL) {
+        /* make sure we haven't generated an unreasonable number of these */
+        int nents = hash_get_num_entries(query_plan_hash);
+        if (nents >= gbl_query_plan_max_queries) {
```

Comment:
Done! I called it max_plan_query_plans since for fingerprints there is max_query_fingerprints to keep a similar formatting, but can change the name if that doesn't make much sense. I also decided to move the whole data structure into the fingerprint one so that the query isn't being copied to two separate places, but still have my old code also